### ISSUE TITLE
NamedTuple or Dataclass instead of misc.Record

### ISSUE THREAD

**Issue by zentarim** [ronf/asyncssh#408]
@ronf hello!

Could you please name the reason why you do not use named tuples or dataclasses to represent all Record-descendants? So far I'm unable to figure it out.

It seems really annoying wrenching out field names from class comments instead of just relying upon IDE-autocompletion.
Not to mention `# pylint: disable=attribute-defined-outside-init` pragmas and comments in the code.

**Comment by ronf** [comment]
Part of the answer to that is historical. The original implementation of the Record subclass (initially added inside the SFTP module) dates back to 2015, before data classes existed in Python. Even today, data classes are only part of the standard library in Python 3.7 and later, and I still have a requirement in AsyncSSH 2.x to support Python 3.6+ (originally, I supported Python 3.4+ in AsyncSSH 1.x).

While there is a back port of data classes for Python 3.6 on PyPI, I'm not sure I want to force that additional external dependency at this point.

As for why I didn't just use a namedtuple here, the Record class was designed to provide a few things that namedtuple didn't at the time:

- Ability to specify default values for each field in the record (though I see now there was an obscure way to do this)
- Ability to specify a formatter to optionally change the representation of field values when converting the record to a string.

Looking more closely at namedtuple, I see that the "typing" module has added a NamedTuple syntax that looks very much like a lightweight data class which could work in Python 3.6, so that's something definitely worth exploring. I'd still need to figure out how to subclass NamedType (which can't be done in the "normal" way) if I wanted to keep my custom string conversion, but this would actually have the benefit of solving another issue around how to add type annotations to my records. I actually ended up having to duplicate the type annotations and my field names in __slots__ in the current approach, which is kind of ugly. Using NamedTuple would look much nicer.

Thanks for the question! I'll definitely follow up on this.

**Comment by zentarim** [comment]
Thanks for the fast and so detailed answer! Glad to see that the project is not abandoned and still evolves.

About dataclasses:
My bad, I forgot that dataclasses is implemented only in 3.7

What about NamedTuple... Yes there are two problems I see:
1) Inheritance from NamedTuple-subclassed classes. I've been messing with it for a while and found no way to make it work.
2) Default values for mutable types (Dict, List, Set, etc). It can't be done. This functional is supported only by dataclasses (via `field` function)

About the print format I have a thought:
I hardly ever need to print Record-descendants as you intended it to be printed. Mostly, I need attributes from there to build up my own logic over it. I agree that Record has a cool formatter, it really has. But would I upset much if it has a NamedTuple or dataclass format? No, I definetely won't. And I don't think that someone would complain much about it.

P.S.
I've got couple of ideas about NamedType and default mutable values. Maybe I can do something about the second problem.

**Comment by ronf** [comment]
I experimented with this a bit more last night, and item 1 (not being able to leverage inheritance) is definitely a show-stopper. While I could find other ways to deal with the logging, there are other methods I want to define on these objects like encode/decode to serialize them in SFTP packet format and from_local to convert between the local OS version of these objects and the SFTP version, and those methods just don't work right as they can't see any of the members of the named tuple.

As for the formatting, that was put in place as part of supporting logging of SFTP requests that used these objects. For example, the current code can produce the following log output:

```
DEBUG:asyncssh.sftp:[conn=0, chan=0] Sending opendir for /tmp
DEBUG:asyncssh.sftp:[conn=0, chan=0] Received handle 00000000
DEBUG:asyncssh.sftp:[conn=0, chan=0] Sending readdir for handle 00000000
DEBUG:asyncssh.sftp:[conn=0, chan=0] Received 3 names
DEBUG:asyncssh.sftp:[conn=0, chan=0]   filename: ., longname: drwxrwxrwt   13 root     wheel         416 Sep 15 06:58 ., attrs: size: 416, uid: 0, gid: 0, permissions: 041777, atime: Wed Sep 15 07:01:41 2021, mtime: Wed Sep 15 06:58:44 2021
DEBUG:asyncssh.sftp:[conn=0, chan=0]   filename: .., longname: drwxr-xr-x    6 root     wheel         192 Aug 22  2019 .., attrs: size: 192, uid: 0, gid: 0, permissions: 040755, atime: Fri Aug 23 01:21:11 2019, mtime: Thu Aug 22 22:48:10 2019
DEBUG:asyncssh.sftp:[conn=0, chan=0]   filename: t.py, longname: -rw-r--r--    1 ronf     wheel        1773 Sep 14 21:51 t.py, attrs: size: 1773, uid: 501, gid: 0, permissions: 100644, atime: Tue Sep 14 21:51:13 2021, mtime: Tue Sep 14 21:51:13 2021
DEBUG:asyncssh.sftp:[conn=0, chan=0] Sending readdir for handle 00000000
DEBUG:asyncssh.sftp:[conn=0, chan=0] Sending close for handle 00000000
DEBUG:asyncssh.sftp:[conn=0, chan=0] Received OK
```

It's also smart enough to only show values that are being changed on something like a chmod call:

```
DEBUG:asyncssh.sftp:[conn=0, chan=0] Sending setstat for /tmp/x, permissions: 000777
DEBUG:asyncssh.sftp:[conn=0, chan=0] Received OK
```

I hadn't considered a potential problem with mutable types. I don't need much of that, but there is a place where one of the fields is currently a list, and another place where it is a dict. It might be possible to make those into read-only mappings/sequences, but I'm not sure if I'd need to do other refactoring to allow that.

That said, I think the change I'm already working on will address some of your concern even without explicitly making the classes into named tuples. Specifically, for type hinting reasons, I've added the fields to each record subclass as type hints. For instance:

```python
class SFTPName(Record):
    """SFTP file name and attributes

       SFTPName is a simple record class with the following fields:

         ========= ================================== ==================
         Field     Description                        Type
         ========= ================================== ==================
         filename  Filename                           `str` or `bytes`
         longname  Expanded form of filename & attrs  `str` or `bytes`
         attrs     File attributes                    :class:`SFTPAttrs`
         ========= ================================== ==================

       A list of these is returned by :meth:`readdir() <SFTPClient.readdir>`
       in :class:`SFTPClient` when retrieving the contents of a directory.

    """

    __slots__ = OrderedDict((('filename', ''), ('longname', ''),
                             ('attrs', SFTPAttrs())))

    filename: BytesOrStr
    longname: BytesOrStr
    attrs: SFTPAttrs
```
I think this might be enough to allow auto-completion of fields, and it also allows for type inferencing when they are referenced. The duplication of these new lines and the previous __slots__ isn't great, but I'm thinking I might be able to adapt some of the code in NamedTuple into my own Record class to allow for the default values to be specified in a NamedTuple-like form, but while still allowing for inheritance. I also ran across http://zecong.hu/2019/08/10/inheritance-for-namedtuples/ while researching this, and that may help to work around some of the inheritance issues. The code is pretty complex, but I may not need a lot of the functionality that was designed to provide.

**Comment by ronf** [comment]
Here's an extended version of the Record type which allows both the types and the default values of record subclasses to be specified in a format similar to NamedTuple, but with a much simpler implementation that allows for proper inheritance, mainly by only allowing the class syntax and not the argument-based syntax that NamedTuple picked up from the older namedtuple class.

The following is a diff that you should be able to apply to the current "develop" branch (or the 2.7.2 release) which adds support for the new Record style (replacing what I previous did of encoding default values in __slots__) and updates all the placed where records are used:

```patch
diff --git a/asyncssh/misc.py b/asyncssh/misc.py
index 3bc9fdf..9e7527f 100644
--- a/asyncssh/misc.py
+++ b/asyncssh/misc.py
@@ -26,9 +26,9 @@ import ipaddress
 import re
 import socket

-from collections import OrderedDict
 from pathlib import Path
 from random import SystemRandom
+from typing import Union

 from .constants import DEFAULT_LANG
 from .constants import DISC_COMPRESSION_ERROR, DISC_CONNECTION_LOST
@@ -39,6 +39,9 @@ from .constants import DISC_PROTOCOL_ERROR, DISC_PROTOCOL_VERSION_NOT_SUPPORTED
 from .constants import DISC_SERVICE_NOT_AVAILABLE


+BytesOrStr = Union[bytes, str]
+
+
 # Define a version of randrange which is based on SystemRandom(), so that
 # we get back numbers suitable for cryptographic use.
 _random = SystemRandom()
@@ -277,10 +280,24 @@ class Options:
         self.prepare(**self.kwargs)


-class Record:
+class _RecordMeta(type):
+    """Metaclass for general-purpose record type"""
+
+    def __new__(mcs, name, bases, ns):
+        if name != 'Record':
+            fields = ns.get('__annotations__', {}).keys()
+            defaults = {k: ns.get(k) for k in fields}
+
+            ns = {k: v for k, v in ns.items() if k not in fields}
+            ns['__slots__'] = defaults
+
+        return super().__new__(mcs, name, bases, ns)
+
+
+class Record(metaclass=_RecordMeta):
     """General-purpose record type with fixed set of fields"""

-    __slots__ = OrderedDict()
+    __slots__ = {}

     def __init__(self, *args, **kwargs):
         for k, v in self.__slots__.items():
diff --git a/asyncssh/process.py b/asyncssh/process.py
index ef2ebf0..75d2c05 100644
--- a/asyncssh/process.py
+++ b/asyncssh/process.py
@@ -23,16 +23,16 @@
 import asyncio
 from asyncio.subprocess import DEVNULL, PIPE, STDOUT
 import codecs
-from collections import OrderedDict
 import inspect
 import io
 import os
 from pathlib import PurePath
 import socket
 import stat
+from typing import Mapping, Optional, Tuple

 from .constants import DEFAULT_LANG, EXTENDED_DATA_STDERR
-from .misc import Error, ProtocolError, Record, open_file
+from .misc import BytesOrStr, Error, ProtocolError, Record, open_file
 from .stream import SSHClientStreamSession, SSHServerStreamSession
 from .stream import SSHReader, SSHWriter

@@ -525,7 +525,7 @@ class SSHCompletedProcess(Record):
          ============ ======================================= =================
          Field        Description                             Type
          ============ ======================================= =================
-         env          The environment the client requested    `str` or `None`
+         env          The environment the client requested    `dict` or `None`
                       to be set for the process
          command      The command the client requested the    `str` or `None`
                       process to execute (if any)
@@ -550,10 +550,14 @@ class SSHCompletedProcess(Record):

     """

-    __slots__ = OrderedDict((('env', None), ('command', None),
-                             ('subsystem', None), ('exit_status', None),
-                             ('exit_signal', None), ('returncode', None),
-                             ('stdout', None), ('stderr', None)))
+    env: Optional[Mapping[str, str]]
+    command: Optional[str]
+    subsystem: Optional[str]
+    exit_status: Optional[int]
+    exit_signal: Optional[Tuple[str, bool, str, str]]
+    returncode: Optional[int]
+    stdout: Optional[BytesOrStr]
+    stderr: Optional[BytesOrStr]


 class SSHProcess:
diff --git a/asyncssh/sftp.py b/asyncssh/sftp.py
index 7c6015a..536cd48 100644
--- a/asyncssh/sftp.py
+++ b/asyncssh/sftp.py
@@ -23,7 +23,6 @@
 """SFTP handlers"""

 import asyncio
-from collections import OrderedDict
 import errno
 from fnmatch import fnmatch
 import inspect
@@ -34,6 +33,7 @@ import posixpath
 import stat
 import sys
 import time
+from typing import Optional, Sequence, Tuple

 from .constants import DEFAULT_LANG

@@ -56,8 +56,8 @@ from .constants import FX_OK, FX_EOF, FX_NO_SUCH_FILE, FX_PERMISSION_DENIED
 from .constants import FX_FAILURE, FX_BAD_MESSAGE, FX_NO_CONNECTION
 from .constants import FX_CONNECTION_LOST, FX_OP_UNSUPPORTED

-from .misc import Error, Record, async_context_manager, get_symbol_names
-from .misc import hide_empty, plural, to_hex
+from .misc import BytesOrStr, Error, Record, async_context_manager
+from .misc import get_symbol_names, hide_empty, plural, to_hex

 from .packet import Byte, String, UInt32, UInt64, PacketDecodeError
 from .packet import SSHPacket, SSHPacketLogger
@@ -797,13 +797,14 @@ class SFTPAttrs(Record):

     """

-    # Unfortunately, pylint can't handle attributes defined with setattr
-    # pylint: disable=attribute-defined-outside-init
-
-    __slots__ = OrderedDict((('size', None), ('uid', None), ('gid', None),
-                             ('permissions', None), ('atime', None),
-                             ('mtime', None), ('nlink', None),
-                             ('extended', [])))
+    size: Optional[int]
+    uid: Optional[int]
+    gid: Optional[int]
+    permissions: Optional[int]
+    atime: Optional[int]
+    mtime: Optional[int]
+    nlink: Optional[int]
+    extended: Sequence[Tuple[bytes, bytes]] = ()

     def _format(self, k, v):
         """Convert attributes to more readable values"""
@@ -915,13 +916,17 @@ class SFTPVFSAttrs(Record):

     """

-    # Unfortunately, pylint can't handle attributes defined with setattr
-    # pylint: disable=attribute-defined-outside-init
-
-    __slots__ = OrderedDict((('bsize', 0), ('frsize', 0), ('blocks', 0),
-                             ('bfree', 0), ('bavail', 0), ('files', 0),
-                             ('ffree', 0), ('favail', 0), ('fsid', 0),
-                             ('flags', 0), ('namemax', 0)))
+    bsize: int = 0
+    frsize: int = 0
+    blocks: int = 0
+    bfree: int = 0
+    bavail: int = 0
+    files: int = 0
+    ffree: int = 0
+    favail: int = 0
+    fsid: int = 0
+    flags: int = 0
+    namemax: int = 0

     def encode(self):
         """Encode SFTP statvfs attributes as bytes in an SSH packet"""
@@ -981,8 +986,9 @@ class SFTPName(Record):

     """

-    __slots__ = OrderedDict((('filename', ''), ('longname', ''),
-                             ('attrs', SFTPAttrs())))
+    filename: BytesOrStr = ''
+    longname: BytesOrStr = ''
+    attrs: SFTPAttrs = SFTPAttrs()

     def _format(self, k, v):
         """Convert name fields to more readable values"""
```

If you could give this a try and see if it addresses the issue you saw with auto-completion in your IDE, I'd really appreciate it! It doesn't actually rely on NamedTuple (which as we've discussed causes problems), but I'm hoping that the IDE might pick up the annotations and the class members from the class objects themselves, and those will be filled in now.

As mentioned above, I had already begun to switch over to something like this in the process of adding type annotations to the code, but with this change I can do both the type annotation and default value of each field in one place, which is much cleaner.

**Comment by zentarim** [comment]
At least, Pycharm understands that hints well.
But with this patch many self-tests fail. I think that you should fill `Optional` fields like this:
```
    size: Optional[int] = None
    uid: Optional[int] = None
    gid: Optional[int] = None
    permissions: Optional[int] = None
    atime: Optional[int] = None
    mtime: Optional[int] = None
    nlink: Optional[int] = None
```
P.S. Event that does not fix them entirely.

**Comment by zentarim** [comment]
By the way, you can use `typing.AnyStr` insted of `Union[str, bytes]`:
https://docs.python.org/3/library/typing.html#typing.AnyStr

**Comment by ronf** [comment]
I'm glad to hear that the types are being picked up...

Which tests are failing for you? I ran the complete set of unit tests here on the patch above and they all passed, testing against Python 3.6 through 3.10. The new Record class is designed to automatically fill in a default value of `None` if a value isn't specified, so the change you suggested should be a no-op.

Regarding AnyStr, it isn't the same as Union[bytes, str]. AnyStr is defined as a type variable, which means that it needs to be bound to a single variable for all uses of that type in a particular scope for it to work. In the local branch where I've added type hints, I do use AnyStr in some places, but there's a need to use the plain Union in others, in places where this is no associated type var to bind to, or where there might be multiple fields where not all of them are one type or the other. I'll look at it more closely when I move this code back to where I have the full set of type hints in place, but the whole bytes vs. str thing has been a real problem for me in doing this type hinting, since in most cases the type cannot be determined statically. It depends on something like the value of an "encoding" argument, which can vary at runtime.

**Comment by zentarim** [comment]
@ronf, my apologies for a late answer!

`=== 1209 passed, 67 skipped in 372.51s (0:06:12) ====`

All is ok, I've just messed up with patch. Hope to see it in the `master` branch ASAP!

**Comment by ronf** [comment]
Thanks for the confirmation -- I'll try to get this into "develop" soon, and then into the next release.

**Comment by ronf** [comment]
This is now in the "develop" branch as commit d143231.

**Comment by zentarim** [comment]
Thanks!
Waiting for 2.7.3

**Comment by ronf** [comment]
Sorry for the delay. This is now released in AsyncSSH 2.8.0.
